// @tma.js/init-data-node/dist/entries/node.js
import * as E from "fp-ts/Either";
import { pipe } from "fp-ts/lib/function.js";
import { createHmac as createHmac$1 } from "node:crypto";
import { h as hashToken$1, s as signFp$1, a as signDataFp$1, v as validateFp$1, b as bufferToArrayBuffer } from "./parsing-B9llLYuO.js";
import { A, E as E2, H, S, c, i, e, p, d, f, g } from "./parsing-B9llLYuO.js";
import { deepSnakeToCamelObjKeys } from "@tma.js/toolkit";
function textToBuffer(text) {
  return Buffer.from(typeof text === "string" ? text : new Uint8Array(text));
}
const createHmac = (data, key) => {
  return bufferToArrayBuffer(
    createHmac$1("sha256", textToBuffer(key)).update(textToBuffer(data)).digest()
  );
};
function hashToken(token) {
  return Buffer.from(hashToken$1(token, createHmac));
}
function isValid(value, token, options) {
  return pipe(
    validateFp(value, token, options),
    E.match(() => false, () => true)
  );
}
function signFp(data, key, authDate, options) {
  return signFp$1(data, key, authDate, signDataFp, options);
}
function sign(data, key, authDate, options) {
  return pipe(
    signFp(data, key, authDate, options),
    E.match((e2) => {
      throw e2;
    }, (v) => v)
  );
}
function signDataFp(data, key, options) {
  return signDataFp$1(false, data, key, createHmac, options);
}
function signData(data, key, options) {
  return pipe(
    signDataFp(data, key, options),
    E.match((e2) => {
      throw e2;
    }, (v) => v)
  );
}
function validateFp(value, token, options) {
  return validateFp$1(false, value, token, signDataFp, options);
}
function validate(value, token, options) {
  pipe(
    validateFp(value, token, options),
    E.mapLeft((error) => {
      throw error;
    })
  );
}
export {
  A as AuthDateInvalidError,
  E2 as ExpiredError,
  H as HexStringLengthInvalidError,
  S as SignatureInvalidError,
  c as SignatureMissingError,
  deepSnakeToCamelObjKeys,
  hashToken,
  isValid,
  i as isValid3rd,
  e as isValid3rdFp,
  p as parse,
  d as parseFp,
  sign,
  signData,
  signDataFp,
  signFp,
  validate,
  f as validate3rd,
  g as validate3rdFp,
  validateFp
};
